### PR TITLE
Kubevirt provider: Use PVCs to back machines

### PR DESCRIPTION
**Pull request by moadqassem**
**What this PR does / why we need it**:
The machine-controller can create Kubevirt Nodes that use a PV as root disk.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #595

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -143,3 +143,7 @@ required = [
 [[constraint]]
   name = "github.com/docker/distribution"
   version = "v2.7.1"
+
+[[constraint]]
```
Comment (alvaroaleman) [inline Gopkg.toml]: I am confused, this doesn't seem to be used anywhere. But the presubmit that checks the dependencies isn't failing either..?

File: examples/kubevirt-local-provisioner.yaml [hunk 2]
```diff
@@ -0,0 +1 @@
...
metadata:
  name: local-storage-provisioner-node-binding
subjects:
  - kind: ServiceAccount
    name: local-storage-admin
    namespace: kube-system
roleRef:
  kind: ClusterRole
  name: local-storage-provisioner-node-clusterrole
  apiGroup: rbac.authorization.k8s.io^M
```
Comment (alvaroaleman) [inline examples/kubevirt-local-provisioner.yaml]: That doesn't look right :)

Comment (alvaroaleman) [reply examples/kubevirt-local-provisioner.yaml]: Also can we extend the kubevirt CR to not do software virtualization? Or is that the default anyways?

File: test/e2e/provisioning/testdata/machinedeployment-kubevirt.yaml [hunk 3]
```diff
@@ -1,3 +1,25 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: kubevirt-config
+  namespace: kubevirt
+  labels:
+    kubevirt.io: ""
+data:
+  feature-gates: "DataVolumes"
+---
```
Comment (alvaroaleman) [inline test/e2e/provisioning/testdata/machinedeployment-kubevirt.yaml]: Please remove the configmap and the PV

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 4]
```diff
@@ -20,6 +20,7 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	v1alpha12 "kubevirt.io/containerized-data-importer/pkg/apis/core/v1alpha1"
```
Comment (alvaroaleman) [inline pkg/cloudprovider/provider/kubevirt/provider.go]: I am confused, this doesn't seem to be used anywhre? Also please rename to `cdiv1alpha1` or something the like

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 5]
```diff
@@ -261,62 +269,99 @@ func (p *provider) Create(machine *v1alpha1.Machine, _ *cloudprovidertypes.Provi
 		return nil, err
 	}
 
-	quantity, err := resource.ParseQuantity("2Gi")
+	quantity, err := resource.ParseQuantity("10Gi")
```
Comment (alvaroaleman) [inline pkg/cloudprovider/provider/kubevirt/provider.go]: Instead of hardcoding this, can we add this to the config, similarily how we did id for the `sourceURL`?
Also please extend the `Validate` if the value of the field can be parsed by `resource.ParseQuantity`

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 6]
```diff
@@ -261,62 +269,99 @@ func (p *provider) Create(machine *v1alpha1.Machine, _ *cloudprovidertypes.Provi
...
 		},
-		Spec: kubevirtv1.VirtualMachineInstanceSpec{
-			Domain: kubevirtv1.DomainSpec{
-				Devices: kubevirtv1.Devices{
-					Disks: []kubevirtv1.Disk{
-						{
-							Name:       "containerdisk",
-							DiskDevice: kubevirtv1.DiskDevice{Disk: &kubevirtv1.DiskTarget{Bus: "virtio"}},
+		Spec: kubevirtv1.VirtualMachineSpec{
+			Running: &running,
```
Comment (alvaroaleman) [inline pkg/cloudprovider/provider/kubevirt/provider.go]: Please use the helpers in `utilpointer "k8s.io/utils/pointer"` for this kind of problem

File: Gopkg.toml (hunk 1 above)
Comment (moadqassem) [reply Gopkg.toml]: I honestly didn't change that manually, the dep cli did, so not quite sure either

File: examples/kubevirt-local-provisioner.yaml (hunk 2 above)
Comment (moadqassem) [reply examples/kubevirt-local-provisioner.yaml]: I think those were added by my IDE somehow, but tbh I was planning on cleaning the PR afterward because at this time I just want it to pass ;) that's why I didn't request a review

File: pkg/cloudprovider/provider/kubevirt/provider.go (hunk 4 above)
Comment (moadqassem) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: How come? it is used: https://github.com/kubermatic/machine-controller/blob/3235-machinecontroller-kubevirt-provider-allow-using-pvs-as-backing-store/pkg/cloudprovider/provider/kubevirt/provider.go#L347-L371

File: pkg/cloudprovider/provider/kubevirt/provider.go (hunk 5 above)
Comment (moadqassem) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: Yeah that's a good idea actually, will do ;)

File: pkg/cloudprovider/provider/kubevirt/provider.go (hunk 6 above)
Comment (moadqassem) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: What problem exactly?

File: examples/kubevirt-local-provisioner.yaml (hunk 2 above)
Comment (moadqassem) [reply examples/kubevirt-local-provisioner.yaml]: Regarding the default virtualization, it's disabled by default

File: pkg/cloudprovider/provider/kubevirt/provider.go (hunk 4 above)
Comment (alvaroaleman) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: ah okay, then just rename, the current import alias is really bad :P

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 7]
```diff
@@ -261,62 +279,93 @@ func (p *provider) Create(machine *v1alpha1.Machine, _ *cloudprovidertypes.Provi
...
 					},
-				},
-				{
-					Name: "emptydisk",
-					VolumeSource: kubevirtv1.VolumeSource{
-						EmptyDisk: &kubevirtv1.EmptyDiskSource{
-							Capacity: quantity,
+					Spec: v1alpha12.DataVolumeSpec{
+						PVC: &corev1.PersistentVolumeClaimSpec{
+							StorageClassName: &storageClassName,
```
Comment (alvaroaleman) [inline pkg/cloudprovider/provider/kubevirt/provider.go]: Please use `utilpointer` here as well

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 8]
```diff
@@ -59,6 +61,8 @@ type RawConfig struct {
 	Memory        providerconfig.ConfigVarString `json:"memory"`
 	RegistryImage providerconfig.ConfigVarString `json:"registryImage"`
```
[the remark this answers is not in the document]
Comment (moadqassem) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: Well, its only added as a label so I think yes

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 9]
```diff
@@ -261,62 +272,91 @@ func (p *provider) Create(machine *v1alpha1.Machine, _ *cloudprovidertypes.Provi
...
 					},
-				},
-				{
-					Name: "emptydisk",
-					VolumeSource: kubevirtv1.VolumeSource{
-						EmptyDisk: &kubevirtv1.EmptyDiskSource{
-							Capacity: quantity,
+					Spec: v1alpha12.DataVolumeSpec{
+						PVC: &corev1.PersistentVolumeClaimSpec{
+							StorageClassName: utilpointer.StringPtr(storageClassName),
```
Comment (alvaroaleman) [inline pkg/cloudprovider/provider/kubevirt/provider.go]: This must be configurable as well

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 10]
```diff
@@ -54,19 +56,21 @@ func New(configVarResolver *providerconfig.ConfigVarResolver) cloudprovidertypes
...
-	CPUs          providerconfig.ConfigVarString `json:"cpus"`
-	Memory        providerconfig.ConfigVarString `json:"memory"`
-	RegistryImage providerconfig.ConfigVarString `json:"registryImage"`
-	Namespace     providerconfig.ConfigVarString `json:"namespace"`
+	Config     providerconfig.ConfigVarString `json:"config,omitempty"`
+	CPUs       providerconfig.ConfigVarString `json:"cpus"`
+	Memory     providerconfig.ConfigVarString `json:"memory"`
+	Namespace  providerconfig.ConfigVarString `json:"namespace"`
+	SourceURL  providerconfig.ConfigVarString `json:"sourceURL"`
+	PVCStorage providerconfig.ConfigVarString `json:"pvcStorage"`
```
Comment (alvaroaleman) [inline pkg/cloudprovider/provider/kubevirt/provider.go]: Can we rename this to  `PVCSize`?

File: test/e2e/provisioning/helper.go [hunk 11]
```diff
@@ -155,13 +155,13 @@ func testScenario(t *testing.T, testCase scenario, cloudProvider string, testPar
 func buildScenarios() []scenario {
 	var all []scenario
 	for _, version := range versions {
-		for _, os := range operatingSystems {
+		for _, ops := range operatingSystems {
```
Comment (alvaroaleman) [inline test/e2e/provisioning/helper.go]: Why did you rename this?

Comment (moadqassem) [reply test/e2e/provisioning/helper.go]: because a package called `os` is called:
https://github.com/kubermatic/machine-controller/pull/595/files/72ab586e9bcce062034ea5ceb3fdb7b3a2d24f16#diff-4b71c77de88f8c32fb99c651cc08ab1fL21

Comment (alvaroaleman) [reply test/e2e/provisioning/helper.go]: That link doesn't work for me. If the thing shadows something, can you just rename the var to `operatingSystem`? The current name could also be random-generated

Comment (moadqassem) [reply test/e2e/provisioning/helper.go]: well in go, generated vars are actually started with `_`, and its a derived variable from the slice `operatingSystems`

File: pkg/cloudprovider/provider/kubevirt/provider.go (hunk 10 above)
Comment (moadqassem) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: Yeah that makes more sense, will do

File: pkg/cloudprovider/provider/kubevirt/provider.go (hunk 9 above)
Comment (moadqassem) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: Agree 👍

File: test/e2e/provisioning/helper.go (hunk 11 above)
Comment (alvaroaleman) [reply test/e2e/provisioning/helper.go]: What exactly is generated about this var? This name doesn't reflect at all what it is, please write readable code.

Comment (moadqassem) [reply test/e2e/provisioning/helper.go]: There is nothing generated about it, you said `The current name could also be random-generated`, so I told you generated vars in go usually starts with  `_`, and to be honest the code is really readable, the slice is called `operatingSystem`, and `ops` is a part of that slice, what is not reliable about  ???

**Comment by alvaroaleman** [comment]
@moadqassem Can you also update `examples/kubevirt-machinedeployment.yaml` and remove `examples/kubevirt-0.10.0.yaml`?

File: pkg/cloudprovider/provider/kubevirt/provider.go [hunk 12]
```diff
@@ -159,12 +176,19 @@ func (p *provider) Get(machine *v1alpha1.Machine, _ *cloudprovidertypes.Provider
...
+			return nil, fmt.Errorf("failed to get VirtualMachine %s: %v", machine.Name, err)
+		}
+		return nil, cloudprovidererrors.ErrInstanceNotFound
+	}
+
 	virtualMachineInstance := &kubevirtv1.VirtualMachineInstance{}
 	if err := sigClient.Get(ctx, types.NamespacedName{Namespace: c.Namespace, Name: machine.Name}, virtualMachineInstance); err != nil {
 		if !kerrors.IsNotFound(err) {
-			return nil, fmt.Errorf("failed to get VirtualMachineInstance %s: %v", machine.Name, err)
+			return &kubeVirtServer{}, nil
```
Comment (alvaroaleman) [inline pkg/cloudprovider/provider/kubevirt/provider.go]: Shouldn't we return the error here? Why should we swallow errors that are not 404?

IMHO we should return the empty kubevirtServer and a nil error if the error _is_ a 404

Comment (moadqassem) [reply pkg/cloudprovider/provider/kubevirt/provider.go]: I fully agree, my bad
